### ISSUE TITLE
psql:src/scratch/cockroach/test.sql:61: ERROR:  could not decorrelate subquery with CASE statement

### ISSUE THREAD

**Issue by markddrake** [cockroachdb/cockroach#85696]
**Describe the problem**

Query returns psql:src/scratch/cockroach/test.sql:61: ERROR:  could not decorrelate subquery. Same query runs in Postgres 14.

**To Reproduce**

Run the following script using the latest Cockroach Docker image

```
--
select version();
--
drop schema if exists t1 cascade;
--
create schema t1;
--
create table t1.test (
  "CountryID"  integer,
  "Border"     geography
);
--
select "CountryID",
       case 
         when ST_GeometryType("Border"::geometry) = 'ST_MultiPolygon' then
         ST_ASEWKB(( 
		   select concat('MULTIPOLYGON(',string_agg("POLYGON",','),')')::geometry
             from (
               select 1 as "X", concat('(',string_agg("RING",','),')') as "POLYGON"
			     from (
                   select level1, level2, concat('(',string_agg(concat(
				     case 
					   when (point ->> 0)::numeric = -180 then
					     '180'
					   else
					     point ->> 0
					 end
					 ,' ',
					 case
					   when (point ->> 1)::numeric = -180 then
					     '180'
					   else
					     point ->> 1
					 end
                     ),','),')') as "RING" 
	                 from JSONB_ARRAY_ELEMENTS(ST_ASGEOJSON("Border")::JSONB -> 'coordinates') WITH ORDINALITY AS t1 (polygon, level1), 
		                  JSONB_ARRAY_ELEMENTS(polygon) WITH ORDINALITY AS t2 (ring,  level2),
		                  JSONB_ARRAY_ELEMENTS(ring) WITH ORDINALITY AS t3 (point, level3)
                    group by  level1, level2
                 ) as "RINGS"
		        group by level1
			 ) as "POLYGONS"
            group by "X"
	     ))        
         when ST_GeometryType("Border"::geometry) = 'ST_Polygon' then
         ST_ASEWKB((
		   select concat('POLYGON(',string_agg("RING",','),')')::geometry
            from (
              select 1 as "X", level1, concat('(',string_agg(concat(point ->> 0,' ',point ->> 1),','),')') as "RING"
	            from JSONB_ARRAY_ELEMENTS(ST_ASGEOJSON("Border")::JSONB -> 'coordinates')  WITH ORDINALITY AS t1 (ring, level1), 
		             JSONB_ARRAY_ELEMENTS(ring)  WITH ORDINALITY AS t2 (point, level2)
                group by  level1
            ) as "RINGS"
            group by "X"
	     ))         
		 when ST_GeometryType("Border"::geometry) = 'ST_Point' then
		   ST_ASEWKB(concat('POINT(',round(ST_X("Border"::geometry)::numeric,18)::text,' ',round(ST_Y("Border"::geometry)::numeric,18)::text,')')::geometry)
		 else
		   ST_ASEWKB("Border"::geometry)
	   end "Border"
 from t1.test limit 10;
--
\q
```

Running the script generates the following

```
C:\Development\YADAMU>psql -Uroot -hyadamu-db2 -d yadamu -p 26257 -a -f src\scratch\cockroach\test.sql
--
select version();
                                       version
-------------------------------------------------------------------------------------
 CockroachDB CCL v22.1.5 (x86_64-pc-linux-gnu, built 2022/07/28 14:58:04, go1.17.11)
(1 row)


--
drop schema if exists t1 cascade;
DROP SCHEMA
--
create schema t1;
CREATE SCHEMA
--
create table t1.test (
  "CountryID"  integer,
  "Border"     geography
);
CREATE TABLE
--
select "CountryID",
       case
         when ST_GeometryType("Border"::geometry) = 'ST_MultiPolygon' then
         ST_ASEWKB((
                   select concat('MULTIPOLYGON(',string_agg("POLYGON",','),')')::geometry
             from (
               select 1 as "X", concat('(',string_agg("RING",','),')') as "POLYGON"
                             from (
                   select level1, level2, concat('(',string_agg(concat(
                                     case
                                           when (point ->> 0)::numeric = -180 then
                                             '180'
                                           else
                                             point ->> 0
                                         end
                                         ,' ',
                                         case
                                           when (point ->> 1)::numeric = -180 then
                                             '180'
                                           else
                                             point ->> 1
                                         end
                     ),','),')') as "RING"
                         from JSONB_ARRAY_ELEMENTS(ST_ASGEOJSON("Border")::JSONB -> 'coordinates') WITH ORDINALITY AS t1 (polygon, level1),
                                  JSONB_ARRAY_ELEMENTS(polygon) WITH ORDINALITY AS t2 (ring,  level2),
                                  JSONB_ARRAY_ELEMENTS(ring) WITH ORDINALITY AS t3 (point, level3)
                    group by  level1, level2
                 ) as "RINGS"
                        group by level1
                         ) as "POLYGONS"
            group by "X"
             ))
         when ST_GeometryType("Border"::geometry) = 'ST_Polygon' then
         ST_ASEWKB((
                   select concat('POLYGON(',string_agg("RING",','),')')::geometry
            from (
              select 1 as "X", level1, concat('(',string_agg(concat(point ->> 0,' ',point ->> 1),','),')') as "RING"
                    from JSONB_ARRAY_ELEMENTS(ST_ASGEOJSON("Border")::JSONB -> 'coordinates')  WITH ORDINALITY AS t1 (ring, level1),
                             JSONB_ARRAY_ELEMENTS(ring)  WITH ORDINALITY AS t2 (point, level2)
                group by  level1
            ) as "RINGS"
            group by "X"
             ))
                 when ST_GeometryType("Border"::geometry) = 'ST_Point' then
                   ST_ASEWKB(concat('POINT(',round(ST_X("Border"::geometry)::numeric,18)::text,' ',round(ST_Y("Border"::geometry)::numeric,18)::text,')')::geometry)
                 else
                   ST_ASEWKB("Border"::geometry)
           end "Border"
 from t1.test limit 10;
psql:src/scratch/cockroach/test.sql:61: ERROR:  could not decorrelate subquery
--
\q

C:\Development\YADAMU>
```
**Expected behavior**
Query should return no rows, or rows if data is pe

Running same against Postrgres 14 Docker image generates 

```

C:\Development\YADAMU>psql -Upostgres -hyadamu-db2 -d yadamu -a -f src\scratch\cockroach\test.sql
Password for user postgres:
--
select version();
                                                           version
-----------------------------------------------------------------------------------------------------------------------------
 PostgreSQL 14.4 (Debian 14.4-1.pgdg110+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 10.2.1-6) 10.2.1 20210110, 64-bit
(1 row)


--
drop schema if exists t1 cascade;
psql:src/scratch/cockroach/test.sql:4: NOTICE:  drop cascades to table t1.test
DROP SCHEMA
--
create schema t1;
CREATE SCHEMA
--
create table t1.test (
  "CountryID"  integer,
  "Border"     geography
);
CREATE TABLE
--
select "CountryID",
       case
         when ST_GeometryType("Border"::geometry) = 'ST_MultiPolygon' then
         ST_ASEWKB((
                   select concat('MULTIPOLYGON(',string_agg("POLYGON",','),')')::geometry
             from (
               select 1 as "X", concat('(',string_agg("RING",','),')') as "POLYGON"
                             from (
                   select level1, level2, concat('(',string_agg(concat(
                                     case
                                           when (point ->> 0)::numeric = -180 then
                                             '180'
                                           else
                                             point ->> 0
                                         end
                                         ,' ',
                                         case
                                           when (point ->> 1)::numeric = -180 then
                                             '180'
                                           else
                                             point ->> 1
                                         end
                     ),','),')') as "RING"
                         from JSONB_ARRAY_ELEMENTS(ST_ASGEOJSON("Border")::JSONB -> 'coordinates') WITH ORDINALITY AS t1 (polygon, level1),
                                  JSONB_ARRAY_ELEMENTS(polygon) WITH ORDINALITY AS t2 (ring,  level2),
                                  JSONB_ARRAY_ELEMENTS(ring) WITH ORDINALITY AS t3 (point, level3)
                    group by  level1, level2
                 ) as "RINGS"
                        group by level1
                         ) as "POLYGONS"
            group by "X"
             ))
         when ST_GeometryType("Border"::geometry) = 'ST_Polygon' then
         ST_ASEWKB((
                   select concat('POLYGON(',string_agg("RING",','),')')::geometry
            from (
              select 1 as "X", level1, concat('(',string_agg(concat(point ->> 0,' ',point ->> 1),','),')') as "RING"
                    from JSONB_ARRAY_ELEMENTS(ST_ASGEOJSON("Border")::JSONB -> 'coordinates')  WITH ORDINALITY AS t1 (ring, level1),
                             JSONB_ARRAY_ELEMENTS(ring)  WITH ORDINALITY AS t2 (point, level2)
                group by  level1
            ) as "RINGS"
            group by "X"
             ))
                 when ST_GeometryType("Border"::geometry) = 'ST_Point' then
                   ST_ASEWKB(concat('POINT(',round(ST_X("Border"::geometry)::numeric,18)::text,' ',round(ST_Y("Border"::geometry)::numeric,18)::text,')')::geometry)
                 else
                   ST_ASEWKB("Border"::geometry)
           end "Border"
 from t1.test limit 10;
 CountryID | Border
-----------+--------
(0 rows)


--
\q

C:\Development\YADAMU>
````

**Environment:**
Cockroach Docker Image

Attempt to work around the problem that st_dumppoints is not supported.

Jira issue: CRDB-18385

**Comment by markddrake** [comment]
The Problem only show up when the case statement is used

The query executes fine without the case

```
yadamu=> select "CountryID",
yadamu->          ST_ASEWKB((
yadamu(>    select concat('POLYGON(',string_agg("RING",','),')')::geometry
yadamu(>             from (
yadamu(>               select 1 as "X", level1, concat('(',string_agg(concat(point ->> 0,' ',point ->> 1),','),')') as "RING"
yadamu(>             from JSONB_ARRAY_ELEMENTS(ST_ASGEOJSON("Border")::JSONB -> 'coordinates')  WITH ORDINALITY AS t1 (ring, level1),
yadamu(>              JSONB_ARRAY_ELEMENTS(ring)  WITH ORDINALITY AS t2 (point, level2)
yadamu(>                 group by  level1
yadamu(>             ) as "RINGS"
yadamu(>             group by "X"
yadamu(>      ))
yadamu-> from t1.test;;
 CountryID | st_asewkb
-----------+-----------
(0 rows)
```

But fails with a simple case 

```
yadamu=>
yadamu=> select "CountryID",
yadamu->        case when ST_GeometryType("Border"::geometry) = 'ST_MultiPolygon' then
yadamu->          ST_ASEWKB((
yadamu(>    select concat('POLYGON(',string_agg("RING",','),')')::geometry
yadamu(>             from (
yadamu(>               select 1 as "X", level1, concat('(',string_agg(concat(point ->> 0,' ',point ->> 1),','),')') as "RING"
yadamu(>             from JSONB_ARRAY_ELEMENTS(ST_ASGEOJSON("Border")::JSONB -> 'coordinates')  WITH ORDINALITY AS t1 (ring, level1),
yadamu(>              JSONB_ARRAY_ELEMENTS(ring)  WITH ORDINALITY AS t2 (point, level2)
yadamu(>                 group by  level1
yadamu(>             ) as "RINGS"
yadamu(>             group by "X"
yadamu(>      ))
yadamu->        else
yadamu->          NULL
yadamu->   end
yadamu-> from t1.test;;
ERROR:  could not decorrelate subquery
yadamu=>
yadamu=>

```

**Comment by markddrake** [comment]
In my case a workaround is to use the COALESCE operator rather than case

```
C:\Development\YADAMU>psql -Uroot -hyadamu-db2 -d yadamu -p 26257 -a -f src\scratch\cockroach\test2.sql
--
select version();
                                       version
-------------------------------------------------------------------------------------
 CockroachDB CCL v22.1.5 (x86_64-pc-linux-gnu, built 2022/07/28 14:58:04, go1.17.11)
(1 row)


--
drop schema if exists t1 cascade;
DROP SCHEMA
--
create schema t1;
CREATE SCHEMA
--
create table t1.test (
  "CountryID"  integer,
  "Border"     geography
);
CREATE TABLE
--
select count(*) from (
select "CountryID",
       COALESCE(
         ST_ASEWKB((
                   select concat('MULTIPOLYGON(',string_agg("POLYGON",','),')')::geometry
             from (
               select 1 as "X", concat('(',string_agg("RING",','),')') as "POLYGON"
                             from (
                   select level1, level2, concat('(',string_agg(concat(
                                     case
                                           when (point ->> 0)::numeric = -180 then
                                             '180'
                                           else
                                             point ->> 0
                                         end
                                         ,' ',
                                         case
                                           when (point ->> 1)::numeric = -180 then
                                             '180'
                                           else
                                             point ->> 1
                                         end
                     ),','),')') as "RING"
                         from JSONB_ARRAY_ELEMENTS(ST_ASGEOJSON("Border")::JSONB -> 'coordinates') WITH ORDINALITY AS t1 (polygon, level1),
                                  JSONB_ARRAY_ELEMENTS(polygon) WITH ORDINALITY AS t2 (ring,  level2),
                                  JSONB_ARRAY_ELEMENTS(ring) WITH ORDINALITY AS t3 (point, level3)
                    group by  level1, level2
                 ) as "RINGS"
                        group by level1
                         ) as "POLYGONS"
            where ST_GeometryType("Border"::geometry) = 'ST_MultiPolygon'
            group by "X"
             )),
         ST_ASEWKB((
                   select concat('POLYGON(',string_agg("RING",','),')')::geometry
            from (
              select 1 as "X", level1, concat('(',string_agg(concat(point ->> 0,' ',point ->> 1),','),')') as "RING"
                    from JSONB_ARRAY_ELEMENTS(ST_ASGEOJSON("Border")::JSONB -> 'coordinates')  WITH ORDINALITY AS t4 (ring, level1),
                             JSONB_ARRAY_ELEMENTS(ring)  WITH ORDINALITY AS t5 (point, level2)
                group by  level1
            ) as "RINGS"
            where ST_GeometryType("Border"::geometry) = 'ST_Polygon'
            group by "X"
             )),
         ST_ASEWKB((
                   select concat('POINT(',round(ST_X("Border"::geometry)::numeric,18)::text,' ',round(ST_Y("Border"::geometry)::numeric,18)::text,')')::geometry
                     from (
                           select "Border"::geometry "Border"
                         )
                        where ST_GeometryType("Border"::geometry) = 'ST_Point'
         )),
                 ST_ASEWKB("Border"::geometry)
           )
 from "WWI_Application"."Countries"
) as "RC";
 count
-------
   190
(1 row)


---
```

**Comment by mgartner** [comment]
This was fixed by #95234, which should be included in v23.1. Closing.
